### ISSUE TITLE
`nix upgrade-nix` does not detect 2.4

### ISSUE THREAD

**Issue by Dzieni** [NixOS/nix#5503]
**Describe the bug**

My collegaue has problems with remote builds using Nix 2.4. To see if it's a matter of different Nix version (I use 2.3 myself), I wanted to upgrade. However, I'm not sure how to do this.

**Steps To Reproduce**

1. Run `nix-channel --update`, then `nix upgrade-nix`.

**Expected behavior**

According to --help, it should `upgrade Nix to the latest stable version`. [Latest stable version is 2.4](https://nixos.org/manual/nix/stable/release-notes/rl-2.4.html), but I get this output instead:

```
[0.0 MiB DL]
replacing old 'nix-2.3.16'
installing 'nix-2.3.16'
upgrade to version 2.3.16 done
```

**`nix-env --version` output**

2.3.16, as mentioned above.

**Additional context**

I also tried to upgrade using `nix-env -u` and it managed to upgrade to `nix-2.4pre-rc1`. Then, after running `nix upgrade-nix`, it downgraded to 2.3:

```
replacing old 'nix-2.4pre-rc1'
installing 'nix-2.3.16'
```

OS: macOS Big Sur 11.3.1

**Comment by Dzieni** [comment]
Is it possibly related to the fact that I run my Nix single-user and it's no longer supported in v2.4?

https://github.com/NixOS/nix/blob/2.4-maintenance/scripts/install-nix-from-closure.sh#L63

If so, I think it should be mentioned in release notes...?

**Comment by abathur** [comment]
> Is it possibly related to the fact that I run my Nix single-user and it's no longer supported in v2.4?

No. Upgrade pulls the latest version from a file in nixpkgs (which still indicates 2.3.16).

I am not sure exactly what the criteria for updating it is, but I assume this is waiting for nixpkgs itself to complete updating (see NixOS/nixpkgs#144197).

**Comment by Dzieni** [comment]
@abathur does it mean that the best workaround for now is to just reinstall Nix from scratch?

**Comment by abathur** [comment]
I would personally treat that as a last resort, if you need exactly the same version. Nixpkgs also includes a nixUnstable attr (also known as nixFlakes) which should be a pre-2.5 build that isn't too far ahead of 2.4 at this point.

I haven't tried using this via nix-env, but if the wiki page for flakes is right, something like `nix-env -iA nixpkgs.nixUnstable` should work.

**Comment by abathur** [comment]
The nixpkgs PR I mentioned before has been merged--this should work now.

**Comment by matu3ba** [comment]
I can confirm that this works now:
```txt
[user@pc zsh]$ nix --version
nix (Nix) 2.4pre20210429_d15a196
[user@pc zsh]$ nix upgrade
error: 'upgrade' is not a recognised command
Try 'nix --help' for more information.
[user@pc zsh]$ nix --help
[user@pc zsh]$ nix-channel --update
unpacking channels...
[user@pc zsh]$ nix upgrade-nix
replacing old 'nix-2.4pre20210429_d15a196'
installing 'nix-2.11.0'
building '/nix/store/lyrzv64mi81lj04fh0qjg6vx5xbqmm6i-user-environment.drv'...
[user@pc zsh]$ nix --version
nix (Nix) 2.11.0
[user@pc zsh]$
```

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
